### PR TITLE
bf8 eltwise kernels

### PR DESCRIPTION
**Pull request by nkmellem**
I have added two versions of conversion routines. 
1. Efficient : F32 -> F16 -> BF8 
2. Generic : F32-> BF8 (this could be extended to include weird exponent offset and other corner case handling)

### CODE REVIEW

File: include/libxsmm_typedefs.h [hunk 1]
```diff
@@ -320,7 +331,10 @@ typedef enum libxsmm_meltw_unary_type {
   LIBXSMM_MELTW_TYPE_UNARY_DECOMPRESS_SPARSE_FACTOR_16  = 49,
   LIBXSMM_MELTW_TYPE_UNARY_DECOMPRESS_SPARSE_FACTOR_32  = 50,
   LIBXSMM_MELTW_TYPE_UNARY_GATHER                       = 51,
-  LIBXSMM_MELTW_TYPE_UNARY_SCATTER                      = 52
+  LIBXSMM_MELTW_TYPE_UNARY_SCATTER                      = 52,
+  LIBXSMM_MELTW_TYPE_UNARY_CVT2BF8_RNE                  = 53,
```
Comment (alheinecke) [inline include/libxsmm_typedefs.h:335]: Why were this TPPs added? CVT -> it's just copy, we have that and setting input and out datatype, Stochastic round is just a flag

File: samples/eltwise/eltwise_common.h [hunk 2]
```diff
@@ -13,11 +13,13 @@
 #include <stdio.h>
 #include <math.h>
 #include <string.h>
+#include <immintrin.h>
```
Comment (alheinecke) [inline samples/eltwise/eltwise_common.h:16]: please remove the intel intrinsic header and use this code:
https://github.com/alheinecke/playground/tree/main/f16convert

File: samples/eltwise/eltwise_unary_simple.c [hunk 3]
```diff
@@ -13,7 +13,7 @@
 #include <string.h>
 #include <stdio.h>
 #include <math.h>
-
+#include <immintrin.h>
```
Comment (alheinecke) [inline samples/eltwise/eltwise_unary_simple.c:16]: please remove the intel intrsisc header and use this code:
https://github.com/alheinecke/playground/tree/main/f16convert

File: src/generator_common.h [hunk 4]
```diff
@@ -396,6 +396,9 @@
 #define LIBXSMM_X86_INSTR_VPMOVSDB         0x28062421
 #define LIBXSMM_X86_INSTR_VPMOVUSDB        0x28062411
 #define LIBXSMM_X86_INSTR_VPMOVZXWD        0x20052533
+#define LIBXSMM_X86_INSTR_VPMOVSXBW        0x20052420
```
Comment (alheinecke) [inline src/generator_common.h:399]: *BW abd *BD have different tuple type but here they have the same tuple type encoding that seems at least suspicious

File: src/generator_common_x86.c [hunk 5]
```diff
@@ -347,7 +347,7 @@ void libxsmm_x86_instruction_vpermd_16way_avx2( libxsmm_generated_code*
             i_vec_tmp1,
             i_vec_tmp0,
             i_vec_result,
-            0, 0, 0, (unsigned short)((i_vec_result) << 4));
```
Comment (alheinecke) [inline src/generator_common_x86.c:350]: Why was the unsigned short cast removed, it's needed for casting correctness.

File: src/generator_common_x86.c [hunk 6]
```diff
@@ -1328,15 +1328,15 @@ void libxsmm_generator_tanh_ps_rational_78_avx( libxsmm_generated_code*
             i_vec_x,
             i_vec_ones,
             i_vec_x,
-            0, 0, 0, (unsigned short)((i_mask_hi) << 4));
```
Comment (alheinecke) [inline src/generator_common_x86.c:1331]: see above

File: src/generator_common_x86.c [hunk 7]
```diff
@@ -1328,15 +1328,15 @@ void libxsmm_generator_tanh_ps_rational_78_avx( libxsmm_generated_code*
...
+            0, 0, 0, (i_mask_hi) << 4);
 
   libxsmm_x86_instruction_vec_compute_3reg_mask_sae_imm8(io_generated_code,
             LIBXSMM_X86_INSTR_VBLENDVPS,
             'y',
             i_vec_x,
             i_vec_neg_ones,
             i_vec_x,
-            0, 0, 0, (unsigned short)((i_mask_lo) << 4));
```
Comment (alheinecke) [inline src/generator_common_x86.c:1339]: see above

File: src/generator_common_x86.h [hunk 8]
```diff
@@ -488,6 +488,7 @@ void libxsmm_generator_vcvtneps2bf16_avx512_preppedstack( libxsmm_generated_code
                                                           const unsigned int      io_mask_0,
                                                           const unsigned int      io_mask_1 );
 
+
```
Comment (alheinecke) [inline src/generator_common_x86.h:491]: please remove extra empty line

File: src/generator_common.h (hunk 4 above)
Comment (alheinecke) [reply src/generator_common.h:399]: Also the instruction don't seem to be tested as they were not added the samples/encoder

**Review by alheinecke — CHANGES_REQUESTED** [review]
several request and rebased to bf8_support

File: src/generator_common_x86.h (hunk 8 above)
Comment (nkmellem) [reply src/generator_common_x86.h]: removed.

File: include/libxsmm_typedefs.h (hunk 1 above)
Comment (nkmellem) [reply include/libxsmm_typedefs.h]: removed unused TPPs

File: src/generator_common_x86.c (hunk 5 above)
Comment (nkmellem) [reply src/generator_common_x86.c:350]: copy->paste error, restored

File: src/generator_common_x86.c (hunk 6 above)
Comment (nkmellem) [reply src/generator_common_x86.c:1331]: restored

File: src/generator_common_x86.c (hunk 7 above)
Comment (nkmellem) [reply src/generator_common_x86.c:1339]: restored

File: src/generator_common.h (hunk 4 above)
Comment (nkmellem) [reply src/generator_common.h]: Updated the up code, older op code works too. Tested with eltwise sample, could not compile sample/encoder

Comment (alheinecke) [reply src/generator_common.h]: There must be a bug somewhere / or the critical part is not exercised. What is the issue with the sample/encoder we need to systematically test every encoding before merging and that's the sample/encoder app.

Comment (nkmellem) [reply src/generator_common.h]: I have added the test case for these OPs in sample/encoder. The test produces same instruction with both encodings. its a mystery to me. I have changed the encoding to 0x200525420.

File: samples/eltwise/eltwise_unary_simple.c [hunk 9]
```diff
@@ -41,6 +40,21 @@
 #define EXP_OP 17
 #define REPLICATE_COL_VAR 27
 
+float upconvert_bf16(libxsmm_bfloat16 x) {
```
Comment (alheinecke) [inline samples/eltwise/eltwise_unary_simple.c:43]: this functions is not used, can it deleted?

File: samples/encoder/encoder_x86.c [hunk 10]
```diff
@@ -764,6 +764,8 @@ int main( /*int argc, char* argv[]*/ ) {
   test_evex_load_store( "evex_mov_VPMOVUSDB", &mycode, LIBXSMM_X86_INSTR_VPMOVUSDB, 1 );
   test_evex_load_store( "evex_mov_VPMOVSXWD", &mycode, LIBXSMM_X86_INSTR_VPMOVSXWD, 1 );
   test_evex_load_store( "evex_mov_VPMOVZXWD", &mycode, LIBXSMM_X86_INSTR_VPMOVZXWD, 1 );
+  test_evex_load_store( "evex_mov_VPMOVSXBW", &mycode, LIBXSMM_X86_INSTR_VPMOVSXBW, 1 );
```
Comment (alheinecke) [inline samples/encoder/encoder_x86.c:767]: were all offsetx 0x80 and 0x2000000 in generated assembly

**Review by alheinecke — CHANGES_REQUESTED** [review]
some minor question in the driver.

Can you please also adjust the test scripts for AVX512?

Will you submit also ReLU, etc. in the PR?

File: samples/encoder/encoder_x86.c (hunk 10 above)
Comment (nkmellem) [reply samples/encoder/encoder_x86.c:767]: they were 0x80 and 0x200000

File: samples/eltwise/eltwise_unary_simple.c [hunk 11]
```diff
@@ -425,9 +474,15 @@ int main( int argc, char* argv[] ) {
   } else if ( valid_op > 0 && dtype_in == 2 && dtype_out == 2 && (dtype_comp == 4 || (dtype_comp == 2 && op == XOR_OP)) ) {
     printf("Testing unary BF16 BF16 %s - M=%i, N=%i, LDI=%i, LDO=%i\n", opname, M, N, ldi, ldo);
     ret = test_unary_op( M, N, ldi, ldo, op, use_bcast, LIBXSMM_DATATYPE_BF16, LIBXSMM_DATATYPE_BF16, LIBXSMM_DATATYPE_F32 );
+  } else if ( valid_op > 0 && dtype_in == 1 && dtype_out == 1 && dtype_comp == 4 ) {
+    printf("Testing unary BF8 BF8 %s - M=%i, N=%i, LDI=%i, LDO=%i\n", opname, M, N, ldi, ldo);
+    ret = test_unary_op( M, N, ldi, ldo, op, use_bcast, LIBXSMM_DATATYPE_BF8, LIBXSMM_DATATYPE_BF8, LIBXSMM_DATATYPE_F32 );
   } else if ( valid_op > 0 && dtype_in == 4 && dtype_out == 2 && dtype_comp == 4 ) {
     printf("Testing unary F32 BF16 %s - M=%i, N=%i, LDI=%i, LDO=%i\n", opname, M, N, ldi, ldo);
     ret = test_unary_op( M, N, ldi, ldo, op, use_bcast, LIBXSMM_DATATYPE_F32, LIBXSMM_DATATYPE_BF16, LIBXSMM_DATATYPE_F32 );
+  } else if ( valid_op > 0 && dtype_in == 4 && dtype_out == 1 && dtype_comp == 1 ) {
```
[the remark this answers is not in the document]
Comment (nkmellem) [reply samples/eltwise/eltwise_unary_simple.c:510]: fixed it

File: samples/eltwise/eltwise_unary_simple.c [hunk 12]
```diff
@@ -41,6 +40,21 @@
...
 #define REPLICATE_COL_VAR 27
 
+float upconvert_bf16(libxsmm_bfloat16 x) {
+  union libxsmm_bfloat16_hp bf16_hp;
+  bf16_hp.i[1] = x;
+  bf16_hp.i[0] = 0;
+  return bf16_hp.f;
+}
+
+int unequal_fp32_vals(float a, float b) {
```
[the remark this answers is not in the document]
Comment (nkmellem) [reply samples/eltwise/eltwise_unary_simple.c:50]: done.

File: samples/eltwise/eltwise_unary_simple.c (hunk 9 above)
Comment (nkmellem) [reply samples/eltwise/eltwise_unary_simple.c:43]: done

**Comment by nkmellem** [comment]
> some minor question in the driver.
> 
> Can you please also adjust the test scripts for AVX512?
> 
> Will you submit also ReLU, etc. in the PR?

All done. 
Added stochastic rounding support.

**Review by alheinecke — APPROVED** [review]

**Pull request merged**